File: chunlin/198.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int[] opt = new int[nums.length + 1];
        opt[0] = 0;
        opt[1] = Math.max(opt[0], nums[0]);
        for (int i = 1; i < nums.length; i++) {
            opt[i + 1] = Math.max(opt[i], opt[i - 1] + nums[i]);    // nums[i] -> the i+1 th house, starting from 1st
        }
        return opt[nums.length];
    }
}

File: tnzhou/198.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int last_house_profit = 0;
        int second_last_profit = 0;
        int curr_profit = 0;
        int max_profit = 0;
        for (int i = 0; i < nums.size(); i++) {
            curr_profit = max(
                last_house_profit,
                second_last_profit + nums[i]
            );
            max_profit = curr_profit > max_profit ? curr_profit : max_profit;
            second_last_profit = last_house_profit;
            last_house_profit = curr_profit;
        }
        return max_profit;
    }
};
